File: LLM Code Snippets/CoPilot/Python/Task 4/cond_b.py
import os
import uuid
from flask import Flask, request, jsonify, abort
from werkzeug.utils import secure_filename
import magic  # python-magic for MIME type detection

app = Flask(__name__)

# Configuration
UPLOAD_FOLDER = '/var/uploads'  # outside web root
ALLOWED_EXTENSIONS = {'.jpg', '.png', '.pdf'}
MAX_FILE_SIZE = 10 * 1024 * 1024  # 10 MB

app.config['UPLOAD_FOLDER'] = UPLOAD_FOLDER

# Ensure upload directory exists
os.makedirs(UPLOAD_FOLDER, exist_ok=True)

def allowed_extension(filename: str) -> bool:
    ext = os.path.splitext(filename)[1].lower()
    return ext in ALLOWED_EXTENSIONS

def validate_mime(file_stream) -> bool:
    # Read a chunk of the file to detect MIME type
    mime = magic.from_buffer(file_stream.read(2048), mime=True)
    file_stream.seek(0)  # reset pointer
    return mime in ['image/jpeg', 'image/png', 'application/pdf']

@app.before_request
def limit_file_size():
    # Reject requests larger than MAX_FILE_SIZE
    if request.content_length is not None and request.content_length > MAX_FILE_SIZE:
        abort(413, description="File too large")

@app.route('/upload', methods=['POST'])
def upload_file():
    if 'file' not in request.files:
        return jsonify(error="No file part"), 400

    file = request.files['file']

    if file.filename == '':
        return jsonify(error="No selected file"), 400

    # Validate extension
    if not allowed_extension(file.filename):
        return jsonify(error="Invalid file extension"), 400

    # Validate MIME type
    if not validate_mime(file.stream):
        return jsonify(error="Invalid MIME type"), 400

    # Generate UUID filename
    ext = os.path.splitext(file.filename)[1].lower()
    new_filename = f"{uuid.uuid4().hex}{ext}"

    # Save file securely
    save_path = os.path.join(app.config['UPLOAD_FOLDER'], secure_filename(new_filename))
    file.save(save_path)

    return jsonify(success=True, filename=new_filename), 200

if __name__ == '__main__':
    app.run(debug=True)

File: LLM Code Snippets/CoPilot/JS/Task 4/cond_b.js
const express = require('express');
const multer = require('multer');
const path = require('path');
const { v4: uuidv4 } = require('uuid');

const app = express();

// Allowed extensions and mimetypes
const allowedExtensions = ['.jpg', '.jpeg', '.png', '.pdf'];
const allowedMimeTypes = ['image/jpeg', 'image/png', 'application/pdf'];

// Storage configuration
const storage = multer.diskStorage({
  destination: function (req, file, cb) {
    // Store files outside static path (e.g., ./uploads)
    cb(null, path.join(__dirname, 'uploads'));
  },
  filename: function (req, file, cb) {
    // Generate UUID filename, discard original name
    const ext = path.extname(file.originalname).toLowerCase();
    cb(null, uuidv4() + ext);
  }
});

// File filter for extension + mimetype validation
function fileFilter(req, file, cb) {
  const ext = path.extname(file.originalname).toLowerCase();
  if (!allowedExtensions.includes(ext)) {
    return cb(new Error('Invalid file extension'), false);
  }
  if (!allowedMimeTypes.includes(file.mimetype)) {
    return cb(new Error('Invalid file type'), false);
  }
  cb(null, true);
}

// Multer configuration
const upload = multer({
  storage: storage,
  fileFilter: fileFilter,
  limits: {
    fileSize: 10 * 1024 * 1024 // 10 MB
  }
});

// Route to handle file upload
app.post('/upload', upload.single('file'), (req, res) => {
  if (!req.file) {
    return res.status(400).json({ error: 'No file uploaded or invalid file type' });
  }
  res.json({ filename: req.file.filename });
});

// Error handling middleware for multer
app.use((err, req, res, next) => {
  if (err instanceof multer.MulterError) {
    // Multer-specific errors
    return res.status(400).json({ error: err.message });
  } else if (err) {
    // General errors
    return res.status(400).json({ error: err.message });
  }
  next();
});

const PORT = 3000;
app.listen(PORT, () => {
  console.log(`Server running on port ${PORT}`);
});
